### ISSUE TITLE
Allow type annotations like `Particle | CustomParticle`

### ISSUE THREAD

**Issue by namurphy** [PlasmaPy/PlasmaPy#1320]
A new feature in Python 3.10 is a [new type union operator](https://docs.python.org/3/whatsnew/3.10.html#pep-604-new-type-union-operator) for built-in Python types.  Currently if we want to have a type hint annotation for multiple types, we need to use something like `typing.Union[int, str]`.  This new capability will eventually enable us to write `int | str` instead.  I don't know if we'd be able to implement these changes for built-in types before PlasmaPy's minimum supported version of Python is 3.10, though we might be able to by enabling deferred evaluation of annotations.  

However, we could implement this ourselves for commonly used PlasmaPy classes such as `Particle` and `CustomParticle`.  I think we'd need to specify the `__or__` dunder.  So, this issue is to make `Particle | CustomParticle` return `typing.Union[Particle, CustomParticle]` and `Particle | None` return `typing.Optional[Particle]`.  

One thing we'll have to think about is if it would be worth starting to implement `Particle | CustomParticle` before we can do `int | str`.  The benefit would be a cleaner implementation, but the drawback would be that we'd have multiple notations for type constructs.

**Comment by namurphy** [comment]
From @StanczakDominik: It might be enough to use regularly `from __future__ import annotations` and not need to do this.  

Alternative could be to use string annotations instead.

We'll need to look into this more...

**Comment by namurphy** [comment]
More from today's meeting: if possible, we should avoid implementing a `__or__` dunder if at all possible, since we'd have to do that repeatedly.

**Comment by namurphy** [comment]
It turns out that we can do this on Python 3.10:
```pycon
>>> from plasmapy.particles import CustomParticle, Particle
>>> particle_union = Particle | CustomParticle
>>> type (particle_union)
<class 'types.UnionType'>
```
...so we don't actually have to do anything to enable this more generally, except wait.  For approximately 1.92 years.  😑

**Comment by StanczakDominik** [comment]
I think it might be possible in older versions as well (how old, I'm not sure) via `from __future__ import annotations` in files that would use this.
